File: DAA/Java Codes/BellmanFord.java
//package Practicals;

import java.util.*;

public class BellmanFord {
    public static void main(String[] args) {
        int V = 6;  // Number of vertices
        int S = 0;  // Source vertex

        // Define edges as a 2D array
        int[][] edges = {
                {3, 2, 6},  // edge from 3 to 2 with weight 6
                {5, 3, 1},  // edge from 5 to 3 with weight 1
                {0, 1, 5},  // edge from 0 to 1 with weight 5
                {1, 5, -3}, // edge from 1 to 5 with weight -3
                {1, 2, -2}, // edge from 1 to 2 with weight -2
                {3, 4, -2}, // edge from 3 to 4 with weight -2
                {2, 4, 3}   // edge from 2 to 4 with weight 3
        };

        // Call the Bellman-Ford function
        int[] dist = bellman_ford(V, edges, S);

        // Print the distance from the source to each vertex
        for (int i = 0; i < V; i++) {
            System.out.print(dist[i] + " ");
        }
        System.out.println("");
    }

    // Bellman-Ford function
    static int[] bellman_ford(int V, int[][] edges, int S) {
        int[] dist = new int[V];

        // Initialize distances to a large value
        for (int i = 0; i < V; i++) dist[i] = (int)(1e8);
        dist[S] = 0;  // Distance to source is 0

        // V-1 times relax the edges
        for (int i = 0; i < V - 1; i++) {
            for (int[] edge : edges) {
                int u = edge[0];
                int v = edge[1];
                int wt = edge[2];
                if (dist[u] != 1e8 && dist[u] + wt < dist[v]) {
                    dist[v] = dist[u] + wt;
                }
            }
        }

        // Check for negative weight cycle
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int wt = edge[2];
            if (dist[u] != 1e8 && dist[u] + wt < dist[v]) {
                int[] temp = new int[1];
                temp[0] = -1;  // Return -1 to indicate a negative cycle
                return temp;
            }
        }

        return dist;
    }
}

File: DAA/C++ Codes/bellmanFord.cpp
 #include <iostream>
#include <climits> // For INT_MAX

using namespace std;

// Global variables
int n, m, src;
int edges[100][3]; // Assuming a maximum of 100 edges
int dist[100];     // Assuming a maximum of 100 vertices

// Function to implement the Bellman-Ford algorithm
void bellmanFord() {
    // Step 1: Initialize distances from source to all other vertices as infinite
    for (int i = 1; i <= n; i++) {
        dist[i] = INT_MAX;
    }
    dist[src] = 0; // Distance from source to itself is 0

    // Step 2: Relax all edges n-1 times
    for (int i = 1; i <= n - 1; i++) {
        for (int j = 0; j < m; j++) {
            int u = edges[j][0]; // Source vertex
            int v = edges[j][1]; // Destination vertex
            int wt = edges[j][2]; // Weight of the edge

            // Relaxation step
            if (dist[u] != INT_MAX && dist[u] + wt < dist[v]) {
                dist[v] = dist[u] + wt;
            }
        }
    }

    // Step 3: Check for negative-weight cycles
    bool hasNegativeCycle = false;
    for (int j = 0; j < m; j++) {
        int u = edges[j][0];
        int v = edges[j][1];
        int wt = edges[j][2];
        if (dist[u] != INT_MAX && dist[u] + wt < dist[v]) {
            hasNegativeCycle = true;
            break;
        }
    }

    // Print all distances from the source vertex
    if (!hasNegativeCycle) {
        cout << "Vertex\tDistance from Source (" << src << ")\n";
        for (int i = 1; i <= n; i++) {
            cout << i << "\t\t" << (dist[i] == INT_MAX ? "INF" : to_string(dist[i])) << endl;
        }
    } else {
        cout << "Negative weight cycle detected." << endl;
    }
}

int main() {
    // Input for number of vertices and edges
    cout << "Number of vertices (n): ";
    cin >> n;
    cout << "Number of edges (m): ";
    cin >> m;
    cout << "Source vertex: ";
    cin >> src;

    // Input edges
    cout << "Enter edges (u, v, weight):\n";
    for (int i = 0; i < m; i++) {
        cin >> edges[i][0] >> edges[i][1] >> edges[i][2];
    }

    // Call the Bellman-Ford algorithm
    bellmanFord();

    return 0;
}